### PR TITLE
fix(flat-storage): Handle chain forks gracefully

### PR DESCRIPTION
**Pull request by nikurt**
Also avoid printing WARNings when too many cached deltas are in memory.
Currently the number of cached deltas doesn't matter, only the number of blocks needed to get from a block to flat head is important.

### CODE REVIEW

File: core/store/src/flat/storage.rs [hunk 1]
```diff
@@ -122,6 +122,12 @@ impl FlatStorageInner {
             blocks.len(),
             first_height.map(|height| height - flat_head.height),
         );
+        if blocks.len() >= Self::HOPS_LIMIT {
+            let shard_id = self.shard_uid.shard_id();
+            let flat_head_height = flat_head.height;
+            let cached_deltas = self.deltas.len();
+            warn!(target: "chain", shard_id, flat_head_height, cached_deltas, "Flat storage cached deltas exceeded expected limits");
+        }
```
Comment (Longarithm) [inline core/store/src/flat/storage.rs:130]: No need to check it here, it is covered at least in add_delta -> update_delta_metrics

**Review by Longarithm — APPROVED** [review]

File: core/store/src/flat/storage.rs [hunk 2]
```diff
@@ -641,6 +649,60 @@ mod tests {
         );
     }
 
+    #[test]
+    fn flat_storage_fork() {
+        init_test_logger();
+        // Create a chain with two forks. Set flat head to be at block 0.
+        let num_blocks = 10 as BlockHeight;
+
+        let chain = MockChain::build((0..num_blocks).collect(), |i| {
```
Comment (ppca) [inline core/store/src/flat/storage.rs:658]: why not use MockChain::chain_with_two_forks(num_blocks);?

**Review by ppca — APPROVED** [review]

File: core/store/src/flat/storage.rs [hunk 3]
```diff
@@ -143,12 +149,10 @@ impl FlatStorageInner {
         );
 
         let cached_changes_size_bytes = bytesize::ByteSize(cached_changes_size);
-        if cached_deltas >= Self::CACHED_CHANGES_LIMIT
-            || cached_changes_size_bytes >= Self::CACHED_CHANGES_SIZE_LIMIT
-        {
+        if cached_changes_size_bytes >= Self::CACHED_CHANGES_SIZE_LIMIT {
```
Comment (wacban) [inline core/store/src/flat/storage.rs:152]: How about the other condition?

File: core/store/src/flat/storage.rs [hunk 4]
```diff
@@ -177,8 +181,11 @@ impl FlatStorageInner {
             if new_head == current_flat_head_hash {
                 return Ok(current_flat_head_hash);
             }
-            let metadata =
-                self.deltas.get(&new_head).ok_or(missing_delta_error(&new_head))?.metadata;
+            let metadata = self
+                .deltas
+                .get(&new_head)
+                .ok_or(self.create_block_not_supported_error(&new_head))?
```
Comment (wacban) [inline core/store/src/flat/storage.rs:187]: Can you document it in the method comment? Or anywhere else more appropriate. Treating error types as control flow mechanism is not that typical pattern so it's worth pointing it out. An alternative would be to move that information from the error to the return type instead.

**Review by wacban — APPROVED** [review]
lgtm

**Pull request merged**